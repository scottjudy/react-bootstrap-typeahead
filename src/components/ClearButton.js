// @flow

import cx from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';

import { isSizeLarge, noop } from '../utils';

import { sizeType } from '../propTypes';
import type { EventHandler, KeyboardEventHandler, Size } from '../types';

const propTypes = {
  label: PropTypes.string,
  onClick: PropTypes.func,
  onKeyDown: PropTypes.func,
  size: sizeType,
};

const defaultProps = {
  label: 'Clear',
  onClick: noop,
  onKeyDown: noop,
};

type Props = {
  className?: string,
  label: string,
  onClick: EventHandler<HTMLButtonElement>,
  onKeyDown: KeyboardEventHandler<HTMLButtonElement>,
  size?: Size,
};

/**
 * ClearButton
 *
 * http://getbootstrap.com/css/#helper-classes-close
 */
const ClearButton = ({
  className,
  label,
  onClick,
  onKeyDown,
  size,
  ...props
}: Props) => (
  <button
    {...props}
    aria-label={label}
    className={cx(
      'close',
      'rbt-close',
      {
        'rbt-close-lg': isSizeLarge(size),
      },
      className
    )}
    onClick={(e: SyntheticEvent<HTMLButtonElement>) => {
      e.stopPropagation();
      onClick(e);
    }}
    onKeyDown={(e: SyntheticKeyboardEvent<HTMLButtonElement>) => {
      // Prevent browser from navigating back.
      if (e.key === 'Backspace') {
        e.preventDefault();
      }
      onKeyDown(e);
    }}
    type="button">
    <span aria-hidden="true">&times;</span>
    <span className="sr-only visually-hidden">{label}</span>
  </button>
);

ClearButton.propTypes = propTypes;
ClearButton.defaultProps = defaultProps;

export default ClearButton;
