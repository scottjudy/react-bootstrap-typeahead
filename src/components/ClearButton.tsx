import cx from 'classnames';
import PropTypes from 'prop-types';
import React, {
  HTMLProps,
  KeyboardEvent,
  KeyboardEventHandler,
  MouseEvent,
  MouseEventHandler,
} from 'react';

import type { Size } from '../types';
import { isSizeLarge, noop } from '../utils';
import { sizeType } from '../propTypes';

const propTypes = {
  label: PropTypes.string,
  onClick: PropTypes.func,
  onKeyDown: PropTypes.func,
  size: sizeType,
};

const defaultProps = {
  label: 'Clear',
  onClick: noop,
  onKeyDown: noop,
};

interface ClearButtonProps extends Omit<HTMLProps<HTMLButtonElement>, 'size'> {
  label?: string;
  onClick: MouseEventHandler<HTMLButtonElement>;
  onKeyDown: KeyboardEventHandler<HTMLButtonElement>;
  size?: Size;
}

/**
 * ClearButton
 *
 * http://getbootstrap.com/css/#helper-classes-close
 */
const ClearButton = ({
  className,
  label,
  onClick,
  onKeyDown,
  size,
  ...props
}: ClearButtonProps): JSX.Element => (
  <button
    {...props}
    aria-label={label}
    className={cx(
      'close',
      'rbt-close',
      {
        'rbt-close-lg': isSizeLarge(size),
      },
      className
    )}
    onClick={(e: MouseEvent<HTMLButtonElement>) => {
      e.stopPropagation();
      onClick(e);
    }}
    onKeyDown={(e: KeyboardEvent<HTMLButtonElement>) => {
      // Prevent browser from navigating back.
      if (e.key === 'Backspace') {
        e.preventDefault();
      }
      onKeyDown(e);
    }}
    type="button">
    <span aria-hidden="true">&times;</span>
    <span className="sr-only visually-hidden">{label}</span>
  </button>
);

ClearButton.propTypes = propTypes;
ClearButton.defaultProps = defaultProps;

export default ClearButton;
