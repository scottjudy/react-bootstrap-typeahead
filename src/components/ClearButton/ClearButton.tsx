import cx from 'classnames';
import PropTypes from 'prop-types';
import React, { HTMLProps, KeyboardEvent, MouseEvent } from 'react';

import type { Size } from '../../types';
import { isSizeLarge, isSizeSmall } from '../../utils';

import { sizeType } from '../../propTypes';

const propTypes = {
  label: PropTypes.string,
  onClick: PropTypes.func,
  onKeyDown: PropTypes.func,
  size: sizeType,
};

const defaultProps = {
  label: 'Clear',
};

export interface ClearButtonProps
  extends Omit<HTMLProps<HTMLButtonElement>, 'size'> {
  label: string;
  size?: Size;
}

/**
 * ClearButton
 *
 * http://getbootstrap.com/css/#helper-classes-close
 */
const ClearButton = ({
  className,
  label,
  onClick,
  onKeyDown,
  size,
  ...props
}: ClearButtonProps): JSX.Element => (
  <button
    {...props}
    aria-label={label}
    className={cx(
      'close',
      'btn-close',
      'rbt-close',
      {
        'rbt-close-lg': isSizeLarge(size),
        'rbt-close-sm': isSizeSmall(size),
      },
      className
    )}
    onClick={(e: MouseEvent<HTMLButtonElement>) => {
      e.stopPropagation();
      onClick && onClick(e);
    }}
    onKeyDown={(e: KeyboardEvent<HTMLButtonElement>) => {
      // Prevent browser from navigating back.
      if (e.key === 'Backspace') {
        e.preventDefault();
      }
      onKeyDown && onKeyDown(e);
    }}
    type="button">
    <span aria-hidden="true" className="rbt-close-content">
      &times;
    </span>
    <span className="sr-only visually-hidden">{label}</span>
  </button>
);

ClearButton.propTypes = propTypes;
ClearButton.defaultProps = defaultProps;

export default ClearButton;
