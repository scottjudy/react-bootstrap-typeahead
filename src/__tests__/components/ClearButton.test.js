import { shallow } from 'enzyme';
import React from 'react';

import ClearButton from '../../components/ClearButton.react';

describe('<ClearButton>', () => {
  let button, onClick;

  beforeEach(() => {
    onClick = jest.fn();
    button = shallow(<ClearButton onClick={onClick} />);
  });

  test('renders a default clear button', () => {
    expect(button.type()).toBe('button');
    expect(button.hasClass('close rbt-close')).toBe(true);
  });

  test('renders a large clear button', () => {
    button.setProps({ bsSize: 'large' });
    expect(button.hasClass('rbt-close-lg')).toBe(true);
  });

  test('registers a click', () => {
    button.simulate('click', { stopPropagation: () => {} });
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
