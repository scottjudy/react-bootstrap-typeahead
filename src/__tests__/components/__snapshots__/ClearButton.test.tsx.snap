// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClearButton> renders a snapshot 1`] = `
<button
  aria-label="Clear"
  className="close rbt-close"
  onClick={[Function]}
  onKeyDown={[Function]}
  type="button"
>
  <span
    aria-hidden="true"
  >
    Ã—
  </span>
  <span
    className="sr-only visually-hidden"
  >
    Clear
  </span>
</button>
`;
